snp_measurement/src/stage0.rs

```diff
@@ -57,6 +101,95 @@ impl Stage0Info {
     }
 
+    /// Gets the SEV-SNP specific pages defined in the firmware SEV metadata section entries.
+    pub fn get_snp_pages(&self) -> Vec<SevMetadataPageInfo> {
+        let sev_metadata_content = *self
+            .parse_firmware_guid_table()
+            .get(&SEV_MEATADATA_GUID)
+            .expect("couldn't find SEV metadata entry in GUID table");
+        assert_eq!(
+            sev_metadata_content.len(),
+            size_of::<u32>(),
+            "invalid length for SEV metadata entry"
+        );
+        // We expect the SEV metadata entry in the GUID table to contain only 4 bytes that represent
+        // the 32-bit unsigned little-endian encoding of the reverse offset from the end of the
+        // firmware image to the start of the SEV metadata section.
+        let mut sev_metadata_offset: u32 = 0;
+        sev_metadata_offset
+            .as_bytes_mut()
+            .copy_from_slice(sev_metadata_content);
+        let sev_metadata_offset = sev_metadata_offset as usize;
+        trace!("SEV metadata offset: {}", sev_metadata_offset);
+        assert!(
+            sev_metadata_offset < self.bytes.len(),
+            "invalid SEV metadata offset"
+        );
+        let sev_metadata_header_start = self.bytes.len() - sev_metadata_offset;
+        let sev_metadata_header_end = sev_metadata_header_start + SEV_METADATA_HEADER_SIZE;
+        let header = SevMetadataHeader::parse(
+            &self.bytes[sev_metadata_header_start..sev_metadata_header_end],
+        );
+        trace!("SEV metadata entry count: {}", header.count);
+        let metadata_entries_end = sev_metadata_header_start + header.length as usize;
+        self.bytes[sev_metadata_header_end..metadata_entries_end]
+            .chunks(SEV_METADATA_ENTRY_SIZE)
+            .map(SevMetadataPageInfo::parse)
+            .collect()
+    }
+
+    /// Parses the GUID table from the firmware image as a map.
+    ///
+    /// The GUID (represented as a u128) of each table entry is the key and the associated data of
+    /// the entry is the value.
+    fn parse_firmware_guid_table(&self) -> HashMap<u128, &[u8]> {
+        let content = self.get_guid_table_content();
+        let mut entry_end = content.len();
+        let mut result = HashMap::new();
+        while entry_end > GUID_TABLE_ENTRY_HEADER_SIZE {
+            let GuidTableEntryHeader {
+                guid: entry_guid,
+                size: entry_size,
+            } = GuidTableEntryHeader::parse(&content[..entry_end]);
+            assert!(entry_end >= entry_size as usize, "invalid entry size");
+            let content_end = entry_end - GUID_TABLE_ENTRY_HEADER_SIZE;
+            let content_start = entry_end - entry_size as usize;
+            if let Some(_existing) = result.insert(entry_guid, &content[content_start..content_end])
+            {
+                panic!("duplicate GUID detected in table");
+            }
+            entry_end = content_start;
+        }
+        result
```

Comment:
You should check that there were no bytes left at this point, otherwise the table is malformed.

Reply:
Good point. I will add a check.